### ISSUE TITLE
CS0165 generated from a case where definite assignment seems satisfied.

### ISSUE THREAD

**Issue by nitz** [dotnet/roslyn#57141]
**Version Used**:  `csc 4.0.0-5.21501.19 (8e1779e1)` via `devenv v17.0.0-pre.5.0` and `dotnet 6.0.100-rc.2.21505.57`

**Steps to Reproduce**:

1. `mkdir test && cd test`
2. `dotnet new console`
3. `notepad Program.cs`
4. Replace the template with the code example below.
5. Save and close.
6. `csc Program.cs` (Or `dotnet build`.)
7. Observe errors.

```csharp
bool qux = true;

int foo; // intentionally uninitialized
int bar; // intentionally uninitialized

if (qux)
{
	foo = 4; // foo assigned when theState == State.A
}
else
{
	bar = 2; // bar assigned when theState != State.A
}

if (qux)
{
	bar = 4; // bar assigned when theState == State.A
	System.Console.WriteLine($"{foo}{bar}"); // CS0165: foo not initialized, bar is.
}
else
{
	foo = 2; // foo assigned when theState != State.A
	System.Console.WriteLine($"{bar}{foo}"); // CS0165: bar not initialized, foo is.
}

System.Console.WriteLine($"{foo}{bar}"); // CS0165: foo and bar not initialized.

// Line 26 only generates CS0165 if the WriteLine() calls on Line 18 and/or Line 23 are removed.
// If 18, 23, and 26 are removed, a CS0219 is generated for both `foo` and `bar`.
```

**Expected Behavior**:
No errors expected.

**Actual Behavior**:
CS0165 is generated for lines 18 and 23.

**Additional Information**:

Apologies for the elementary example, I wanted to try and distill what I'm seeing to its purest form.

I'm not super well versed on of the [rules](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#precise-rules-for-determining-definite-assignment) for definite assignment. Based on my reading though, it seems like this case should find `foo` and `bar` definitely assigned by time they're used.

I'm not quite understanding the rules of definite assignment with control flow transfers, but it seems that a case like this could be an edge where the compiler believes it possibly not assigned? With the first assignment being in a branch, and the other assignment being in a separate branch, I could see how the compiler would look at each of those control transfers as "maybe" `foo` gets assigned, and maybe `bar` gets assigned for each. But with the second conditional being the exact same as the first but the opposite variable being assigned, there's no way that the pair of both `foo` and `bar` reach any of the `WriteLine` statements without being assigned.

I took a read around some of the similar issues here, finding #51043 and #56975 being the most similar. I'm unsure if what I'm seeing here falls into the same hat of 'by design' that I should take over to csharplang to discuss, or if it's something closer to a defect.

Thanks for reading!
